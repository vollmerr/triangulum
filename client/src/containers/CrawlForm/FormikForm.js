import React from 'react';
import {
  Dimmer,
  Loader,
  Button,
  Form,
} from 'semantic-ui-react';

import schema from './schema';
import withFormik from './withFormik';

import Field from '../../components/Form/Field';
import ErrorMessage from '../../components/Form/ErrorMessage';

/**
 * Form to wrap with formik HOC
 */
class FormikForm extends React.PureComponent {
  renderFields = () => (
    Object.keys(schema).map(field => (
      <Field key={field} formikProps={this.props} field={schema[field]} />
    ))
  )

  render() {
    const {
      handleSubmit,
      isSubmitting,
    } = this.props;

    const submitProps = {
      fluid: true,
      icon: 'send',
      size: 'large',
      type: 'submit',
      color: 'violet',
      content: 'Send',
      disabled: isSubmitting,
    };

    return (
      <div>
        <Dimmer active={isSubmitting} inverted>
          <Loader>Crawling the interwebs...</Loader>
        </Dimmer>

        <Form size={'large'} onSubmit={handleSubmit}>
          {this.renderFields()}
          <Button {...submitProps} />
        </Form>

        <ErrorMessage schema={schema} {...this.props} />
      </div>
    );
  }
}

export default withFormik(FormikForm);
