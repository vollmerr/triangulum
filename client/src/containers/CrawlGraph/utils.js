/**
 * Stratify function
 *
 * Takes a flat array of node objects containing url and parent url fields, and processes 
 * it into a hierarchical structure, specifically an array containing a single object 
 * with a children array and so on. This mimics d3's stratify function, although is designed 
 * to create the specific formatting requried by react-d3-tree. Additionally, each node
 * object is also given a nodeSvgShape field specified by react-d3-tree for individual node
 * styling. More can be read here: https://github.com/bkrem/react-d3-tree#usage
 *
 * Params: nodes - flat array of nodes objects
 * Returns: data - array with a single object containing children objects
 */
export default function stratify(nodes) {
  
  let data = [];
  findTargetPath(nodes);
  seed(nodes[0], data);
  
  if(nodes.length > 1) {
    for(let i = 0; i < nodes.length; i++) 
      buildHierarchy(nodes[i], data[0]);
  }
  
  addStyling(data[0]);
  return data;

  // Finds target path from the nodes
  function findTargetPath(nodes) {    
    addTargetPathVar(nodes);
    if(nodes[nodes.length - 1].targetFound === 1)
      addTargetPath(nodes, nodes.length - 1, 0);

    // Assigns each object a "targetPath" key and value
    function addTargetPathVar(nodes) {
      for (let i = 0; i < nodes.length; i++)
        nodes[i].targetPath = 0;
    }
    
    // Recursivley determines target path by linear search
    function addTargetPath(nodes, chiIdx, iter) {
      nodes[chiIdx].targetPath = 1;
      if(nodes[chiIdx].parent != null) {
        for(let i = 0; i < nodes.length; i++) {
          if(nodes[chiIdx].parent === nodes[i].url)
            addTargetPath(nodes, i, iter);
        }
      }
    }
  }
  
  // Adds root node to hierarchy
  function seed(node, data) {
    addNodeObj(node, data);
  }

  // Adds new node object to end of array
  function addNodeObj(node, arr) {
    if(node.title) {
      arr.push({ 
        name: node.title,
        data: {
          url: node.url, 
          targetPath: node.targetPath,
          targetFound: node.targetFound,
        },
      });
    }
    else {
      arr.push({ 
        name: node.url,
        data: {
          url: node.url, 
          targetPath: node.targetPath,
          targetFound: node.targetFound,
        },
      });
    }
  }

  // Recursively builds hierarchy 
  function buildHierarchy(node, dataObj) {
    if(dataObj.data.url === node.parent) 
      insertChild(node, dataObj);
    else if(dataObj.children) {
      for(let i = 0; i < dataObj.children.length; i++) 
        buildHierarchy(node, dataObj.children[i]);
    }        
  }

  // Inserts child node obj onto current node
  function insertChild (node, dataObj) {
    if(!dataObj.children) 
      dataObj.children = [];
    addNodeObj(node, dataObj.children);
  }

  // Recursively goes through hierarchy adding styling object
  function addStyling(dataObj) {
    styling(dataObj);
    if(dataObj.children) {
      for(let i = 0; i < dataObj.children.length; i++) 
        addStyling(dataObj.children[i]);
    }        
  }

  // Add nodeSvgShape object to nodes for individual styling
  function styling(dataObj) {

    // Add empty objects
    dataObj.nodeSvgShape = {};
    dataObj.nodeSvgShape.shapeProps = {};

    // Set size
    if(dataObj.data.targetFound) 
      dataObj.nodeSvgShape.shapeProps.r = 20;
    else 
      dataObj.nodeSvgShape.shapeProps.r = 10;

    // Set fill
    if(dataObj.children) {
      if(dataObj.data.targetPath)
        dataObj.nodeSvgShape.shapeProps.fill = "#c0f7cc";
      else
        dataObj.nodeSvgShape.shapeProps.fill = "#e8e8e8";
    } 
    else 
      dataObj.nodeSvgShape.shapeProps.fill = "#fff";

    // Set outline color
    if(dataObj.data.targetPath) {
      if(dataObj.data.targetFound)
        dataObj.nodeSvgShape.shapeProps.stroke = "#dc3545";
      else
        dataObj.nodeSvgShape.shapeProps.stroke = "#28a745";
    }
    else
      dataObj.nodeSvgShape.shapeProps.stroke = "#b7b7b7";

    // Set outline width
    if(dataObj.data.targetFound) 
      dataObj.nodeSvgShape.shapeProps.strokeWidth = 5; 
    else 
      dataObj.nodeSvgShape.shapeProps.strokeWidth = 3;

    // Set cursor when over element to pointer
    dataObj.nodeSvgShape.shapeProps.cursor = 'pointer';
  }
}
