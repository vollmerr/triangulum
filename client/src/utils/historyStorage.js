import uuid from 'uuid/v4';

// NOTE: all the things are stored as strings in localStorage...

export const getItems = () => JSON.parse(localStorage.getItem('crawlHistory'));

export const getItem = ({ id }) => getItems()[id];

export const getCurrent = () => {
  const id = localStorage.getItem('crawlId');
  return id ? getItem({ id }) : {};
};

export const setCurrent = ({ id }) => localStorage.setItem('crawlId', id);

export const newItem = () => {
  const id = uuid();
  const crawlHistory = JSON.stringify({
    ...getItems(),
    [id]: {
      id,
      created: new Date(),
    },
  });
  localStorage.setItem('crawlHistory', crawlHistory);
  localStorage.setItem('crawlId', id);
  return id;
};

export const updateItem = ({ id, ...item }) => {
  const crawlHistory = JSON.stringify({
    ...getItems(),
    [id]: {
      ...getItem({ id }),
      ...item,
      updated: new Date(),
    },
  });
  localStorage.setItem('crawlHistory', crawlHistory);
};

export const reset = () => {
  const crawlHistory = JSON.stringify({});
  localStorage.setItem('crawlHistory', crawlHistory);
  localStorage.setItem('crawlId', '');
};

export const init = () => {
  if (!localStorage.getItem('crawlHistory')) {
    reset();
  }
};
