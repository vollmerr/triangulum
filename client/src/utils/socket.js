const host = process.env.REACT_APP_URL;

let ws = new WebSocket(host); // eslint-disable-line

ws.init = () => {
  ws = new WebSocket(host);
};

ws.onclose = ws.init;

export default ws;
