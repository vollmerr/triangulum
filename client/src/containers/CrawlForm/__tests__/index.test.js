import React from 'react';
import { shallow } from 'enzyme';

import * as historyStorage from '../../../utils/historyStorage';

import FormikForm from '../FormikForm';
import { CrawlForm } from '../index';

const request = { k1: 'v1' };
const response = { k2: 'v2' };
const error = { message: 'test error' };

const props = {
  updateData: jest.fn(),
  history: {
    push: jest.fn(),
  },
};

let wrapper;
let instance;
let id;
describe('CrawlForm', () => {
  beforeEach(() => {
    historyStorage.reset();
    wrapper = shallow(<CrawlForm {...props} />);
    instance = wrapper.instance();
    instance.onRequest({ request });
    id = wrapper.state('id');
  });

  describe('onRequest', () => {
    it('should add a new history item with the request', () => {
      expect(id).toBeDefined();
      expect(historyStorage.getItem({ id }).request).toEqual(request);
    });
  });

  describe('onSuccess', () => {
    beforeEach(() => {
      instance.onSuccess({ response });
    });

    it('should update the graphs data', () => {
      expect(props.updateData).toBeCalledWith({ data: response });
    });

    it('should update the history with the response', () => {
      expect(historyStorage.getItem({ id }).response).toEqual(response);
    });

    it('should navigate to the graph', () => {
      expect(props.history.push).toBeCalledWith('/graph');
    });
  });

  describe('onFailure', () => {
    beforeEach(() => {
      instance.onFailure({ error });
    });

    it('should update the history with an error', () => {
      expect(historyStorage.getItem({ id }).error).toEqual(error);
    });
  });

  describe('render', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render a `FormikForm`', () => {
      expect(wrapper.find(FormikForm).length).toEqual(1);
    });
  });
});
