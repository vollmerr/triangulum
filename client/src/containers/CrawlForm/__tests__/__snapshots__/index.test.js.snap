// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrawlForm should render correctly 1`] = `
<WithFormik(FormikForm)
  onFailure={[Function]}
  onRequest={[Function]}
  onSuccess={[Function]}
/>
`;
